File: app/src/main/java/com/moisesborges/tvaddict/episodes/EpisodesActivity.java
package com.moisesborges.tvaddict.episodes;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import com.moisesborges.tvaddict.App;
import com.moisesborges.tvaddict.R;
import com.moisesborges.tvaddict.adapters.ItemClickListener;
import com.moisesborges.tvaddict.models.Embedded;
import com.moisesborges.tvaddict.models.Episode;
import com.moisesborges.tvaddict.models.Show;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class EpisodesActivity extends AppCompatActivity implements EpisodesView {

    public static final String ARG_SEASON_NUMBER = "com.moisesborges.tvaddict.episodesshow.EpisodesActivity.seasonNumber";
    private static final String ARG_SHOW = "com.moisesborges.tvaddict.episodesshow.EpisodesActivity.show";

    private Show mShow;
    private int mSeasonNumber;

    @BindView(R.id.episodes_recycler_view)
    RecyclerView mRecyclerView;

    @Inject
    EpisodesPresenter mEpisodesPresenter;

    private ItemClickListener<Episode> mEpisodeClickListener =
            episode -> mEpisodesPresenter.changeEpisodeSeenStatus(episode, mShow);

    private EpisodesAdapter mEpisodesAdapter = new EpisodesAdapter(mEpisodeClickListener);

    private Unbinder mUnbinder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_episodes);
        mUnbinder = ButterKnife.bind(this);
        injectDependecies();

        mShow = getIntent().getParcelableExtra(ARG_SHOW);
        mSeasonNumber = getIntent().getIntExtra(ARG_SEASON_NUMBER, 0);

        setupRecyclerView();
    }

    private void injectDependecies() {
        ((App) getApplicationContext()).getAppComponent().inject(this);
    }

    private void setupRecyclerView() {
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setAdapter(mEpisodesAdapter);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mEpisodesPresenter.bindView(this);
        mEpisodesPresenter.loadEpisodes(mSeasonNumber, mShow);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mEpisodesPresenter.unbindView();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mUnbinder.unbind();
    }

    public static void start(@NonNull Context context,
                             @NonNull Show show,
                             int seasonNumber) {
        Intent intent = new Intent(context, EpisodesActivity.class);
        intent.putExtra(ARG_SHOW, show);
        intent.putExtra(ARG_SEASON_NUMBER, seasonNumber);
        context.startActivity(intent);
    }

    @Override
    public void displayEpisodes(@NonNull List<Episode> episodes) {
        mEpisodesAdapter.setEpisodes(episodes);
    }

    @Override
    public void refreshEpisode(@NonNull Episode episode) {
        mEpisodesAdapter.updateEpisode(episode);
    }

    @Override
    public void setShow(@NonNull Show showFromDb) {
        mShow = showFromDb;
    }

    public static class EpisodesAdapter extends RecyclerView.Adapter<EpisodesAdapter.ViewHolder> {

        private final List<Episode> mEpisodes = new ArrayList<>();
        private final ItemClickListener<Episode> mClickListener;

        public EpisodesAdapter(@Nullable ItemClickListener<Episode> clickListener) {
            mClickListener = clickListener;
        }

        @Override
        public EpisodesAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View layout = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_episode, parent, false);
            return new ViewHolder(layout);
        }

        @Override
        public void onBindViewHolder(EpisodesAdapter.ViewHolder holder, int position) {
            Episode episode = mEpisodes.get(position);
            holder.bind(episode);
        }

        @Override
        public int getItemCount() {
            return mEpisodes.size();
        }

        public void setEpisodes(@NonNull List<Episode> episodes) {
            mEpisodes.clear();
            mEpisodes.addAll(episodes);
            notifyDataSetChanged();
        }

        public void updateEpisode(@NonNull Episode episode) {
            int indexOf = mEpisodes.indexOf(episode);
            notifyItemChanged(indexOf);
        }

        public class ViewHolder extends RecyclerView.ViewHolder {

            @BindView(R.id.episode_name_text_view)
            TextView mEpisodeNameTextView;
            @BindView(R.id.watched_check_box)
            CheckBox mWatchedCheckBox;

            public ViewHolder(View itemView) {
                super(itemView);
                ButterKnife.bind(this, itemView);
            }

            @OnClick(R.id.watched_check_box)
            @SuppressWarnings("unused")
            public void onWatchedCheckBoxClick() {
                int position = getAdapterPosition();
                Episode episode = mEpisodes.get(position);
                mClickListener.consume(episode);
            }

            public void bind(@NonNull Episode episode) {
                mEpisodeNameTextView.setText(episode.getName());
                mWatchedCheckBox.setChecked(episode.wasWatched());
            }
        }
    }
}

File: app/src/main/java/com/moisesborges/tvaddict/episodes/EpisodesActivity.kt
package com.moisesborges.tvaddict.episodes

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.Menu
import android.view.MenuItem
import com.moisesborges.tvaddict.App
import com.moisesborges.tvaddict.R
import com.moisesborges.tvaddict.adapters.EpisodesAdapter
import com.moisesborges.tvaddict.adapters.ItemClickListener
import com.moisesborges.tvaddict.models.Episode
import com.moisesborges.tvaddict.models.Show
import kotlinx.android.synthetic.main.activity_episodes.*
import kotlinx.android.synthetic.main.layout_toolbar.*
import javax.inject.Inject

class EpisodesActivity : AppCompatActivity(), EpisodesView {

    private lateinit var show: Show
    private var seasonNumber: Int = 0

    @Inject
    lateinit var episodesPresenter: EpisodesPresenter

    private val toggleEpisodeStatusListener = ItemClickListener<Episode> { episode -> episodesPresenter.changeEpisodeSeenStatus(episode, show) }
    private val openEpisodeDetailsListener = ItemClickListener<Episode> { episode -> episodesPresenter.openEpisodeDetails(episode) }
    private val episodesAdapter = EpisodesAdapter(toggleEpisodeStatusListener, openEpisodeDetailsListener)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_episodes)
        injectDependecies()

        show = intent.getParcelableExtra<Show>(ARG_SHOW)
        seasonNumber = intent.getIntExtra(ARG_SEASON_NUMBER, 0)

        setupRecyclerView()
        setupToolbar()
    }


    private fun injectDependecies() {
        (applicationContext as App).appComponent.inject(this)
    }

    private fun setupRecyclerView() {
        episodes_recycler_view.layoutManager = LinearLayoutManager(this)
        episodes_recycler_view.setHasFixedSize(true)
        episodes_recycler_view.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
        episodes_recycler_view.adapter = episodesAdapter
    }

    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        toolbar.title = show.name
        toolbar.subtitle = resources.getString(R.string.season_number_label, seasonNumber)
    }

    override fun onStart() {
        super.onStart()
        episodesPresenter.bindView(this)
        episodesPresenter.loadEpisodes(seasonNumber, show)
    }

    override fun onStop() {
        super.onStop()
        episodesPresenter.unbindView()
    }

    override fun displayEpisodes(episodes: List<Episode>) {
        episodesAdapter.setEpisodes(episodes)
    }

    override fun refreshEpisode(episode: Episode) {
        episodesAdapter.updateEpisode(episode)
    }

    override fun setShow(showFromDb: Show) {
        show = showFromDb
    }

    override fun showEpisodeDetails(episode: Episode) {
        EpisodeDetailsBottomSheet.show(episode, supportFragmentManager)
    }

    override fun refreshAllEpisodes() {
        episodesAdapter.notifyDataSetChanged()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_episodes, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            android.R.id.home -> {
                close()
            }
            R.id.action_set_all_as_seen -> {
                episodesPresenter.setAllEpisodesStatusAsSeen(seasonNumber, show)
            }
            R.id.action_set_all_as_unseen -> {
                episodesPresenter.setAllEpisodesStatusAsUnseen(seasonNumber, show)
            }
            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
        return true
    }

    private fun close() {
        finish()
    }

    companion object {

        private val ARG_SEASON_NUMBER = "com.moisesborges.tvaddict.episodesshow.EpisodesActivity.seasonNumber"
        private val ARG_SHOW = "com.moisesborges.tvaddict.episodesshow.EpisodesActivity.show"

        @JvmStatic
        fun start(context: Context,
                  show: Show,
                  seasonNumber: Int) {
            val intent = newIntent(context, show, seasonNumber)
            context.startActivity(intent)
        }

        @JvmStatic
        fun newIntent(context: Context, show: Show, seasonNumber: Int): Intent {
            val intent = Intent(context, EpisodesActivity::class.java)
            intent.putExtra(ARG_SHOW, show)
            intent.putExtra(ARG_SEASON_NUMBER, seasonNumber)
            return intent
        }
    }
}
